src/test/FuzzerImpl.cpp

```diff
@@ -97,6 +123,79 @@ makeAssetCode(int digit)
     return code;
 }
 
+void
+generateStoredLedgerKeys(StoredLedgerKeys& keys)
+{
+    size_t constexpr NUM_UNVALIDATED_LEDGER_KEYS = 0x08U;
+    static_assert(NUM_UNVALIDATED_LEDGER_KEYS < NUM_STORED_LEDGER_KEYS,
+                  "No room for valid ledger keys");
+
+    auto const firstUnvalidatedLedgerKey =
+        NUM_STORED_LEDGER_KEYS - NUM_UNVALIDATED_LEDGER_KEYS;
+
+    for (auto index = 0; index < firstUnvalidatedLedgerKey; ++index)
+    {
+        keys[index] =
+            LedgerEntryKey(LedgerTestUtils::generateValidLedgerEntry());
+    }
+
+    for (auto index = firstUnvalidatedLedgerKey; index < NUM_STORED_LEDGER_KEYS;
+         ++index)
+    {
+        size_t const entrySize = 3;
+        keys[index] = autocheck::generator<LedgerKey>()(entrySize);
+    }
+}
+
+void
+storeLedgerKey(StoredLedgerKeys& keys, LedgerKey const& key)
+{
+    auto byte = getShortKey(key);
+    keys[byte % NUM_STORED_LEDGER_KEYS] = key;
```

Comment:
Yes, that's what I meant to do -- I figured that if we unpacked a byte that didn't correspond to a key created by setup, then we'd have a pregenerated one, and if we did, we'd get the one from setup.  I thought that was your intent in suggesting the use of a byte from a hash instead of a counter as an index, but I may have misinterpreted.

Reply:
> You aren't guaranteed to have any invalid/non-existent ledger keys. I don't know if that is occurring here, but that fact that I can't tell is already a problem

I could make `storeLedgerKey()` update `keys[byte % (NUM_STORED_LEDGER_KEYS - NUM_UNVALIDATED_LEDGER_KEYS)]` instead of `keys[byte % NUM_STORED_LEDGER_KEYS] `; that would preserve the (currently 8) unvalidated ledger keys.  Or maybe better, I could drop overwrites of the unvalidated keys by returning early from `storeLedgerKey()` if `byte >= NUM_STORED_LEDGER_KEYS - NUM_UNVALIDATED_LEDGER_KEYS`.  What do you think?

Reply:
> I'm actually curious what the distribution is, so can you share the percentage of keys in the array that are (a) invalid/non-existent and (b) existent of each type?

Stats at present:

- Valid, existing keys: 100
  - Accounts: 34
  - Trustlines: 3
  - Offers: 13
  - Data: 0 (we're not creating any during setup)
  - Claimable balances: 50
- Valid, non-existing keys: 147
  - Accounts: 4
  - Trustlines: 36
  - Offers: 39
  - Data: 37
  - Claimable balances: 31
- Invalid keys: 9 (`TrustLine accountID is issuer`, `XLM TrustLine?`, `TrustLine asset is invalid`)

There are a handful of existing trustlines left, but indeed only a handful -- most of them have been overwritten by offers, I think, because both of those use single bytes of account IDs as their key indexes.

We could make separate arrays for each key type and mix in some keys of the wrong type as well as (pseudo-)randomly invalid ones into each array, but if we do I'd rather post that as a follow-on PR than add more complexity to this one.  We could leave it as it is until then, or we could do some simpler partial amelioration:  for example, we could make `storeLedgerKey()` overwrite ledger keys only a fraction of the time.

Reply:
This is even more interesting than I anticipated. That's an excellent observation about the last byte situation. Let me go think about this.

Reply:
Jon and I discussed this for a while.  One possibility I suggested was just to hash the whole `LedgerKey` (as we do in the entry cache) in `getShortKey()`, instead of switching on the type.  That would somewhat ameliorate the problem of offers almost obliterating trustlines, because they wouldn't have the correlation between their keys anymore.  However, there would still be a bias towards entries created later during setup being better-represented in the stored list.

So now another possibility that occurs to me is that we could populate the stored ledger keys list _after_ executing all the setup operations, using `LedgerTxn::getAllEntries()` at the end of `TransactionFuzzer::initialize()`.  That is, something like:

```
    // Get the list of ledger entries created during setup to place into
    // mStoredLedgerKeys.
    std::vector<LedgerEntry> init, live;
    std::vector<LedgerKey> dead;
    ltxOuter.getAllEntries(init, live, dead);

    // Setup should only create entries; there should be no dead entries, and
    // only one "live" (modified) one:  the root account.
    assert(dead.empty());
    assert(live.size() == 1);
    assert(live[0].data.type() == ACCOUNT);
    assert(live[0].data.account().accountID == root.getPublicKey());

    // Store the ledger entries created during setup in mStoredLedgerKeys.

    auto initEntry = init.cbegin();
    auto storeElem = mStoredLedgerKeys.begin();
    auto const endStore = mStoredLedgerKeys.begin() +
                          FuzzUtils::NUM_STORED_LEDGER_KEYS -
                          FuzzUtils::NUM_UNVALIDATED_LEDGER_KEYS;

    for (; initEntry != init.end() && storeElem != endStore;
         ++initEntry, ++storeElem)
    {
        LOG_TRACE(DEFAULT_LOG, "Setup set stored ledger key entry {} to {}",
                  std::distance(init.cbegin(), initEntry),
                  xdr_to_string(*initEntry));
        *storeElem = LedgerEntryKey(*initEntry);
    }
```

I think that would address Jon's challenge (1) because we'd stop storing keys if we reached the limit of the validated ones, and challenge (2) because all ledger entries would be represented (we only create 160; none would overwrite any other).

That would also mean that we wouldn't have to do anything to store ledger keys while executing the setup operations at all:  `storeLedgerKeyIDsForOp()`, and therefore also `storeLedgerKey()`, could be removed altogether.

(Indeed, the code could even be re-arranged to generate the valid-but-not-setup-generated and the unvalidated ledger keys in the room left after the setup ledger keys had been stored, thus gathering all the ledger-key-generating/storing code into that one place.)

Reply:
Great, this is very much in line with what I was thinking. Two suggestions:
1. Assert if the number of entries to record is more than the space you have available (this way we will know to adopt a different strategy if set-up reaches this threshold)
2. Make more use of the STL to actually transfer the entries

Implement it, and I'll review it.

Reply:
Thanks.  I've pushed a commit to implement it with your suggestions.  I increased the number of unvalidated keys, because with the new scheme, that significantly increased the number of invalid (as opposed to valid non-existing) keys and in particular allowed a greater variety of invalid trustlines (such as a trustline with the same issuer and trustor, and a couple of XLM trustlines).

I also pushed the stronger `MANAGE_OFFER_CREATED` assertion that we discussed.

There are two other intermediate fixes that we discussed as well (avoiding the overwrite of unvalidated keys and using the right offer ID when storing offers), but both are superseded by the commit with the new post-setup-ops mechanism for generating `mStoredLedgerKeys`.

Once the new commit is reviewed, I can add one to gather the generation of invalid entries into the same place as the generation of the valid ones, as I mentioned this change would allow.

Reply:
> Once the new commit is reviewed, I can add one to gather the generation of invalid entries into the same place as the generation of the valid ones, as I mentioned this change would allow.

Actually, I think maybe I shouldn't do that -- leaving `generateStoredLedgerKeys()` as it is allows `OverlayFuzzer` to keep using it.